Initial commits for integration of HPE OneView resources with Ansible

fgbulsoni opened the pull request:
##### SUMMARY
HPE OneView is an infrastructure automation engine built with software intelligence. It streamlines provisioning and lifecycle management across compute, storage and fabric and enables IT staff to control resources programmatically through a unified API.

It is heavily focused on the DevOps workflow and has [many integrations](https://www.hpe.com/us/en/solutions/developers/composable.html) already, including an [ansible module](https://github.com/HewlettPackard/oneview-ansible) not integrated into core.

This PR aims to add the OneView base class to be used for all OneView resources, together with the module for managing the [HPE OneView FC Network resource](http://h17007.www1.hpe.com/docs/enterprise/servers/oneview3.0/cic-api/en/api-docs/current/index.html#rest/fc-networks): `FcNetworkModule` and its unit tests.

This resource has the `present` and `absent` states which will manage all of its aspects, including creating and updating the resources (using `present`) and removal of resources (using `absent`).

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
`FcNetworkModule`

##### ANSIBLE VERSION
```
ansible 2.4.0 (devel de4a8b83df) last updated 2017/06/22 18:57:00 (GMT +000)
  config file = None
  configured module search path = ['/home/vagrant/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/vagrant/temp/ansible/lib/ansible
  executable location = /home/vagrant/temp/ansible/bin/ansible
  python version = 3.5.3 (default, Jun 22 2017, 17:14:20) [GCC 4.8.4]
```


##### ADDITIONAL INFORMATION
Example of usage:
```yaml
    - name: Create a Fibre Channel Network
      oneview_fc_network:
        config: "{{ config }}"
        state: present
        validate_etag: false
        data:
          name: 'Test FC Network'
          fabricType: 'FabricAttach'
          linkStabilityTime: '30'
          autoLoginRedistribution: True
      delegate_to: localhost
      register: fc_network_1

    - name: Update the Fibre Channel Network changing the attribute autoLoginRedistribution to True
      oneview_fc_network:
        config: "{{ config }}"
        state: present
        data:
          name: 'Test FC Network'
          autoLoginRedistribution: True
          fabricType: 'FabricAttach'
          linkStabilityTime: '30'
      delegate_to: localhost

    - name: Delete the Fibre Channel Network
      oneview_fc_network:
        config: "{{ config }}"
        state: absent
        data: "{{ fc_network_1.ansible_facts.fc_network }}"
      delegate_to: localhost
      register: deleted
```

fgbulsoni committed:
Initial commit for integration of HPE OneView resources with Ansible Core. Adding FC Network and FC Network Fact modules and unit tests, and OneView base class for all OV resources.

mattclay commented:
@fgbulsoni Have you considered mocking `hpOneView` instead of importing it? That way we're only testing the module, rather than its dependencies.

Here's another PR which is doing that: https://github.com/ansible/ansible/pull/25450

fgbulsoni commented:
> Have you considered mocking hpOneView instead of importing it?

Hey @mattclay , thanks for the feedback! I'll look into the PR mentioned and see if I can do the same. :octocat:

fgbulsoni committed:
Adding mocks for oneview module imports and removing shebangs from files that offended CI rules.

robinro commented:
I don't see how this relates to me.

fgbulsoni commented:
> I don't see how this relates to me.

Hey @robinro , yeah, not sure what made the bot cc you. Thank you for replying though!

@mattclay , I've added mocks to the `hpOneView` module and now it seems we're hitting an error related to the `collections` method `OrderedDict` which we call inside the [oneview](https://github.com/fgbulsoni/ansible/blob/fb9701b5533266a5d10ca584b5b8fe77019a7123/lib/ansible/module_utils/oneview.py) module_utils file.

Our module does make use of the `OrderedDict` functionality, but I'm not sure about copying the entire method from `collections` and mocking it for it to work on `Python 2.6`, as it will be quite big and I don't feel this will add to the tests.

Would you have any suggestion on how to deal with this? Skipping the 2.6 tests perhaps?

fgbulsoni committed:
Changing OrderedDict imports and calls syntax for unit tests to pass

fgbulsoni commented:
I've fixed the unit tests issues regarding OrderedDict and the unit tests for 2.6 should be passing now.

It seems that for some reason I can't pinpoint the `other` tests are failing on this command `pip install --disable-pip-version-check -c test/runner/requirements/constraints.txt -r test/runner/requirements/sanity.txt coverage junit-xml`.

From the logs and running it locally, it appears to fail when trying to install yamllint with this output:
`Collecting yamllint
/usr/local/lib/python2.6/site-packages/pip/_vendor/requests/packages/urllib3/util/ssl_.py:318: SNIMissingWarning: An HTTPS request has been made, but the SNI (Subject Name Indication) extension to TLS is not available on this platform. This may cause the server to present an incorrect TLS certificate, which can cause validation failures. You can upgrade to a newer version of Python to solve this. For more information, see https://urllib3.readthedocs.io/en/latest/security.html#snimissingwarning.
  SNIMissingWarning
/usr/local/lib/python2.6/site-packages/pip/_vendor/requests/packages/urllib3/util/ssl_.py:122: InsecurePlatformWarning: A true SSLContext object is not available. This prevents urllib3 from configuring SSL appropriately and may cause certain SSL connections to fail. You can upgrade to a newer version of Python to solve this. For more information, see https://urllib3.readthedocs.io/en/latest/security.html#insecureplatformwarning.
  InsecurePlatformWarning
  Using cached yamllint-1.8.0-py2.py3-none-any.whl
Collecting pathspec (from yamllint)
  Using cached pathspec-0.5.2.tar.gz
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-build-omI9Qg/pathspec/setup.py", line 6, in <module>
        from pathspec import __author__, __email__, __license__, __project__, __version__
      File "pathspec/__init__.py", line 33, in <module>
        from .pathspec import PathSpec
      File "pathspec/pathspec.py", line 9, in <module>
        from . import util
      File "pathspec/util.py", line 167
        return {normalize_file(path, separators=separators): path for path in files}
                                                                    ^
    SyntaxError: invalid syntax
`

I've not added yamllint nor am I sure why is it failing, this looks to me like code that was already in the repo. Any help/ideas here?

mattclay commented:
@fgbulsoni I'm working on a fix now.

mattclay commented:
@fgbulsoni The issue has been resolved. I've restarted CI for this PR.

mattclay commented:
These failures are also unrelated. They've been fixed and I'll restart CI again.

fgbulsoni commented:
> These failures are also unrelated. They've been fixed and I'll restart CI again.

Many thanks for that @mattclay ! 

Is there anything else that needs to be addressed in this PR or may I wait for it to be merged?

Best regards,
Felipe.

fgbulsoni commented:
@mattclay and community, Hello again! 

A question was raised by a member of our team and I'd like to hear your/the community thoughts about it.

The "OneView" solution is not *actually* a `Cloud` solution, it is a means of delivering infrastructure as code, so it should make real infrastructure management as easy as dealing with cloud infrastructure. 

This has received a new term and is currently being called [Composable Infrastructure](https://www.hpe.com/us/en/what-is/composable-infrastructure.html).

Taking that into consideration: Would it be possible to have the module under a `composable_infrastructure` or `composable` folder instead of `cloud` inside `/lib/ansible/modules`? 

Or would that be something not currently supported/accepted?

fgbulsoni committed:
Changing import of unittest to use ansible.compat.tests

fgbulsoni commented:
After rebasing to address the ansible changes to tox requirements, the only test that seems to be failing here looks to be the 'TEST=freebsd/10.3-STABLE', with the following error:
```
2017-07-03 11:45:13 ERROR: Timeout waiting for freebsd/10.3-STABLE instance 3c07bd31-eba3-42e9-bee7-cdc44b671e9d.
```
@mattclay Am I correct to assume this is a build issue? Is there anything I should do about it?

Regards,

fgbulsoni commented:
> @fgbulsoni this PR contains more than one new module.

Weird, I'm submitting the fc_network_facts and the fc_network modules in this PR, but it was my understanding that a facts module for a resource should be submitted together with the module for managing the resource.

Is this not correct? The fc_network module requires the fc_network_facts to work as well, and this error had not appeared before.

Can someone give me some inputs on this? Also, on the failing build and questions asked? Is the bot broken ?

gundalow commented:
@fgbulsoni Hi, 
The check for more than one module has just been added to `ansibot` today.

> The fc_network module requires the fc_network_facts to work as well, and this error had not appeared before.
Could you explain that a bit more, that doesn't sounds right. Modules shouldn't talk to other modules.

Shared code goes into `module_utils`
http://docs.ansible.com/ansible/dev_guide/developing_modules_in_groups.html#your-first-pull-request explains the directory structure a bit more

gundalow commented on lib/ansible/modules/cloud/hpe/oneview_fc_network.py:
```diff
@@ -0,0 +1,121 @@
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+ANSIBLE_METADATA = {'metadata_version': '1.0',
+                    'status': ['stableinterface'],
```
preview

gundalow commented on lib/ansible/modules/cloud/hpe/oneview_fc_network_facts.py:
```diff
@@ -0,0 +1,108 @@
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+ANSIBLE_METADATA = {'status': ['stableinterface'],
```
preview

fgbulsoni committed:
Changing status on ansible metadata from stableinterface to preview

fgbulsoni commented:
> Could you explain that a bit more, that doesn't sounds right. Modules shouldn't talk to other modules.
> Shared code goes into module_utils

Hello @gundalow , Thank you for replying! I've fixed the PR according to your comments.

Let me try to clarify some of my questions:

All shared code is already inside the `module_utils`.
However, these two modules (`FcNetworkModule` and `FcNetworkFactsModule`) are related to the same resource, `OneView FC Networks`.

The `FcNetworkFactsModule` module contains all the GET calls to the resource, any calls that do not modify the resource but just query information are contained in this module.

The `FcNetworkModule` module contains all of the operations that may modify a `OneView FC Network` resource, so it is able to create/update/remove those resources and enforce a `state`.

Since they both relate to the same resource, but one is a query module and the other is a management module, I was under the impression that they could be sent together in order to have a full view of the resource capabilities.

Should I be sending the `FcNetworkFactsModule` and  `FcNetworkModule` modules in different PRs even though they are so closely related?

mattclay commented:
I've removed the broken flag on this PR so the bot can continue processing it.

fgbulsoni commented:
> I've removed the broken flag on this PR so the bot can continue processing it.

@mattclay thank you for that.

Could I get a clarification on what direction to move to now?
Is it OK to leave both modules in this PR as they relate to the same resource?
Should I send the `FcNetworkModule` & the `FcNetworkFactsModule` in separate PRs? If so, is there a preference of which I should be sending in first?

Regards,

fgbulsoni committed:
Removed FcNetworkFacts module so the PR features only one module

fgbulsoni commented:
@mattclay @gundalow I've removed the `FcNetworkFactsModule` module that was in this PR previously, so it now only contains the base class for the OneView modules, the `FcNetworkModule` and the unit tests for the module.

I think this is ready_for_review, please let me know if there is something else I should change in the PR or if it can be considered ready to merge.

Regards,

mattclay commented:
CI failure in unit tests:

```
E       NameError: global name 'importlib' is not defined

test/units/modules/cloud/hpe/hpe_test_utils.py:77: NameError
```

fgbulsoni commented:
> CI failure in unit tests:

@mattclay Hey Matt, thanks for the review. 

After we talked I tried removing the hpOneView from the requirements list entirely, which caused the importlib issue to be hit on 2.6.

I then got acquainted with the `__import__` method and was able to properly remove the importlib and hpOneView imports from the tests.

The CI tests are all green now, please let me know if you see anything else that catches your eye in this PR.

mattclay commented:
@fgbulsoni The most recent CI failure is from a test that was just added to cover existing requirements from the [module checklist](http://docs.ansible.com/ansible/dev_guide/developing_modules_checklist.html):

Import custom packages in try/except and handled with fail_json() in main() e.g.

```python
try:
    import foo
    HAS_LIB=True
except:
    HAS_LIB=False
```

fgbulsoni committed:
Removed hpOneView requirement from unit tests and fixed importlib issues for 2.6 compatibility

fgbulsoni commented:
> The most recent CI failure is from a test that was just added to cover existing requirements from the module checklist

Thanks for the update Matt, will make the modifications so that the tests pass then

fgbulsoni committed:
Skipping future import in case it is not available

fgbulsoni commented:
Hey @mattclay , I've made the changes requested to the PR. Please review again :octocat:

fgbulsoni commented:
@mattclay @gundalow If possible, there are two things I'm still uncertain of and would appreciate clarification:

1-  Is it possible to have the HPE OneView modules under a new folder inside the `lib/ansible/modules/` to represent that it is [Composable Infrastructure](https://www.hpe.com/us/en/what-is/composable-infrastructure.html) ? Something like  `lib/ansible/modules/composable` or  `lib/ansible/modules/composable_infrastructure`, since those would be more accurate than `lib/ansible/modules/cloud`.

2- Is it OK that the `author` of the module is different from the person submitting it? Will the submitter be the maintainer for the module? 
I ask this because the FcNetworkModule was originally developed by someone no longer working in HPE, and that person no longer maintains it, so I'd need to be the maintainer for this module.

Could I get a clarification on these two questions and also know if there is anything else I need to modify in the PR?

Regards,

gundalow commented:
@fgbulsoni Good questions :)
1) I'd say `lib/ansible/modules/remote_management/oneview/` would be the best place for these

2) If the modules were developed at HPE, it's fine for you to put `(c) HPE` at the top of the file and list the maintainer under the `author:` field

mattclay commented:
The pep8 failure above is an unrelated issue which has been fixed. Re-running CI.

mattclay commented:
CI failure in python 3.x unit tests:

```
ImportError while importing test module '/root/src/github.com/ansible/ansible/test/units/modules/remote_management/hpe/test_oneview_fc_network.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
test/units/modules/remote_management/hpe/test_oneview_fc_network.py:19: in <module>
    from oneview_module_loader import FcNetworkModule
E   ImportError: No module named 'oneview_module_loader'
```

fgbulsoni committed:
Moving module from cloud to remote_management and changing author to match the current maintainer of the code

fgbulsoni commented:
Thanks for the comments @gundalow & @mattclay , I've modified the PR so that I'm listed as the author and I've changed the location of the module to be under `remote_management` as suggested.
Please let me know if there's something else I should change in this PR. :octocat:

fgbulsoni commented:
Guys, could I get a review here? I believe all comments have been addressed and we're currently passing with the CI. 
The PR is ready to be merged if no other changes are requested.

fgbulsoni commented:
ready_for_review

fgbulsoni commented:
bot_status

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,782 @@
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
```
Should be BSD license for module_utils rather than GPLv3+ (take a look at basic.py, for instance, for proper license).

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,782 @@
+from __future__ import (absolute_import,
+                        division,
+                        print_function,
+                        unicode_literals)
```
Do not use unicode_literals.  It makes it harder to read code rather than easier (because not every file uses unicode_literals and then you have to first check the top of each file for whether it's present and then context switch mentally.)

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,782 @@
+                        unicode_literals)
+
+try:
+    from future import standard_library
```
I think we should avoid usage of the future library.  It tends to override the way python works globally which could cause problems for other libraries.

abadger replied:
Note: We do have a copy of the six library bundled in ansible module_utils so you can use that to aid in py2/py3 compatibility:
``` python
from ansible.module_utils import six
```

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,782 @@
+import os
+import collections
+from ansible.module_utils.basic import AnsibleModule
+from copy import deepcopy
```
We like to separate out our imports as stdlib imports, stdlib from imports, blank line, 3rd party libs, blank line, ansible.module_utils libs.  So it will look like this:
```
import collections
import json
import logging
import os
from copy import deepcopy

from ansible.module_utils.basic import AnsibleModule
```

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,782 @@
+if HAS_FUTURE:
+    standard_library.install_aliases()
+
+logger = logging.getLogger(__name__)
```
We usually use the logging facilities built into AnsibleModule instead of the stdlib's logging module.
```
module = AnsibleModule([...])
module.log()
```

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,782 @@
+            validate_etag_support (bool): Enables support to eTag validation.
+        """
+
+        argument_spec = self.__build_argument_spec(additional_arg_spec, validate_etag_support)
```
Note: We never use the double underscore private convention.  We just use the single underscore private convention.

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,782 @@
+            self.module.exit_json(**result)
+
+        except HPOneViewException as exception:
+            self.module.fail_json(msg='; '.join(str(e) for e in exception.args))
```
catchall exception handling is generally not a good idea as it neither gives a better user message nor aids in debugging.  If you feel you must do it this way, be sure to return the stacktrace (retrieve that using raceback.format_exc() and put it into the exception parameter to fail_json).

abadger replied:
Also note, str can traceback if e contains non-ascii.  Better to use the to_native function from ansible.module_utils._text.

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,782 @@
+        Returns:
+            A dictionary with the expected arguments for the AnsibleModule.exit_json
+
+        """
```
We generally use sphinx docstrings which are formatted like this:
```
        """
        Generic implementation of the absent state for the OneView resources.

        It checks if the resource needs to be removed.

        :arg dict resource: Resource to delete.
        :arg str method: Function of the OneView client that will be called for resource deletion. Usually delete or remove.
        :return: A dictionary with the expected arguments for :meth:`ansible.module_utils.basic.AnsibleModule.exit_json`
"""
```
http://www.sphinx-doc.org/en/stable/domains.html#info-field-lists

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,782 @@
+        )
+
+    @staticmethod
+    def transform_list_to_dict(list_):
```
Generally, staticmethod is a good indicator that this should be a function at the toplevel of the module_util, not a method of the class.  This looks like a prime example of that.

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,782 @@
+            if isinstance(value, dict):
+                ret.update(value)
+            else:
+                ret[str(value)] = True
```
Use to_native() with the parameter errors='surrogate_or_strict' here.  (the default error handler is surrogate_or_replace.  That is appropriate for informational things as undecodable/encodable values will have a replacement character substituted.  For dictionary keys or other data that you would rather throw an error than risk data corruption, surrogate_or_strict is appropriate.)

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,782 @@
+            logging.basicConfig(level=logging.DEBUG, datefmt='%Y-%m-%d %H:%M:%S',
+                                format='%(asctime)s %(levelname)s %(name)s %(message)s',
+                                filename=LOGFILE, filemode='a')
+        return logger
```
@bcoca what do you think of this?  Should the modules just use AnsibleModule.log() instead?

bcoca replied:
yes, since users really want to control this globally, modules should not implement their own logging, think about no_log violations and remote target syslog controls.

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,782 @@
+            return ret
+
+        for value in list_:
+            if isinstance(value, dict):
```
Note: Most of the time, isinstance should use something from collections instead of the type.  In this instance, that should be:
```
if isinstance(value, Mapping):
```
https://docs.python.org/2/library/collections.html#collections-abstract-base-classes

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,782 @@
+        return logger
+
+
+class ResourceComparator():
```
A quick look suggests that the rest of the classes in this file would be better written as python modules instead.  What I'd do is create a hponeview package and make several python modules inside of there:
```
+ module_utils
|--+ hponeview
    |--+ __init__.py      <= Empty file
    |--+ oneviewmodule.py  <= Contains the HpOneView module and immediate support
    |--+ serverprofile.py  <=  Contains the methods from server profile as functions.  The pieces of the resource classes could also go here or into their own file.
```

It looks like these classes are being used to establish separate namespaces for the resources, not really as objects to hold stateful data.  In Python, namespaces are usually implemented as python modules (the separate files above).  classes are reserved for things that need to be instantiated to hold stateful data.

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,782 @@
+            bool: True when equal, False when different.
+        """
+        resource1 = deepcopy(first_resource)
+        resource2 = deepcopy(second_resource)
```
Is there a reason to do deepcopies here (and at the top of a few other methods)?  deepcopies are slow and it doesn't look like the data is being modified (for which you might want to have a pristine copy and a modified copy) so you can probably discard the deepcopy to improve your speed.

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,782 @@
+            return False
+
+        # Checks all keys in first dict against the second dict
+        for key in resource1.keys():
```
Since you're not modifying the data, you can just do:
```
for key in resource:
```
here.

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,782 @@
+        if not original_list:
+            return updated_list
+
+        items_map = collections.OrderedDict([(i[key], i.copy()) for i in original_list])
```
There's a lot of copying going on here as well.  Both on this line and on line 456.  Copying is slow so you want to eliminate any that aren't needed.

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,782 @@
+
+class ResourceMerger():
+    @staticmethod
+    def merge_list_by_key(original_list, updated_list, key, ignore_when_null=[]):
```
You don't modify ignore_when_null in this function so it's probably harmless to use [] as its default value but it's a bad habit to get into.  You should try to always use a immutable as a default value.  In this case, you can do: ```ignore_when_null=tuple()```.

abadger replied:
If you're unfamiliar with why that is, you should probably google it.  It has to do with python processing the function declaration once when the function is declared and therefore there's only one copy of the default value which is used every time the function is called.  If you have a mutable container as a default value, it will not be recreated between invocations so it may not be empty the second time you call the function.

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,782 @@
+            item_key = item[key]
+            if item_key in items_map:
+                for ignored_key in ignore_when_null:
+                    if ignored_key in item and not item[ignored_key]:
```
You probably want to compare this to None specifically instead of just doing a check of whether ```item[ignored_key]``` is a false value.  Otherwise you'll also catch 0, False, and empty containers here.  Comparing to None would look like this:
```
if ignored_key in item and item[ignored_key] is None:
```

abadger replied:
Was this change not needed?

fgbulsoni replied:
I think I had actually missed it, with github closing the comments and all. Fixed now.

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,782 @@
+                for ignored_key in ignore_when_null:
+                    if ignored_key in item and not item[ignored_key]:
+                        item.pop(ignored_key)
+                merged_items[item_key] = items_map[item_key].copy()
```
This will fail if ```updated_list``` has items which are not present in ```original_list```.  You probably want something like this instead:
``` python
merged_items[item_key] = items_map.get(item_key, {})
merged_items[item_key].update(item)
```

abadger replied:
Scratch this comment, I see that you do conditionalize item_key in items_map above this.

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,782 @@
+            else:
+                merged_items[item_key] = item.copy()
+
+        return [val for (_, val) in merged_items.items()]
```
Better to do this like this:
``` python
return list(merged_items.values())
```

Using list there will make sure that it is a list on python 3 (rather than a DictView).

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,782 @@
+        return [val for (_, val) in merged_items.items()]
+
+
+class SPKeys(object):
```
In python, this is probably better represented as a global dictionary in a separate module (as outlined above).

Something like file serviceprovider.py:

SPKEYS = {
    'ID': 'id',
    'NAME': 'name',
[..]
}
```

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,782 @@
+        return merged_data
+
+    def _merge_san_storage_paths(self, merged_data, resource):
+        existing_volumes_map = collections.OrderedDict([(i[SPKeys.ID], i) for i in resource[
```
It doesn't look like you need an OrderedDict here.  A regular dict will do (and be faster and take up less memory).

abadger commented on lib/ansible/modules/remote_management/hpe/oneview_fc_network.py:
```diff
@@ -0,0 +1,121 @@
+version_added: "2.4"
+requirements:
+    - "python >= 2.7.9"
+    - "hpOneView >= 3.1.0"
```
Perhaps these (the requirements) should be in the documentation fragments?

fgbulsoni replied:
I've moved the python requirement to the doc fragments now, but left the hpOneView since its minimum supported version should be specific to each resource

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,782 @@
+        else:
+            self.oneview_client = OneViewClient.from_json_file(self.module.params['config'])
+
+    def execute_module(self):
```
You could use an abstractbaseclass to enforce the requirement for subclasses to implement execute_module() instead of raising an exception inside of the function.  The advantage of an abc is that it will raise when the class is parsed rather than when the function is called.  https://docs.python.org/2/library/abc.html

abadger commented on lib/ansible/modules/remote_management/hpe/oneview_fc_network.py:
```diff
@@ -0,0 +1,121 @@
+    data:
+        description:
+            - List with the Fibre Channel Network properties.
+        required: true
```
@jimi-c are we okay with modules that just have a free-form data parameter?

abadger replied:
jimi-c said on slack that this is fine.

We have added a new feature in 2.4 that allows specifying the structure of objects inside of a list if you want to do that.  It could also be done as an update to the module.  (I'm not entirely sure how to use it yet, knowledge of usage is still percolating out).

abadger commented on lib/ansible/utils/module_docs_fragments/oneview.py:
```diff
@@ -0,0 +1,61 @@
+      description:
+        - Path to a .json configuration file containing the OneView client configuration.
+          The configuration file is optional. If the file path is not provided, the configuration will be loaded from
+          environment variables.
```
It could be useful to specify that the config file is on the managed machine rather than the controller.  This bites some people sometimes.  Should probably also mention what the environment variables the configuration is loaded from are.

abadger replied:
Or maybe say look in the notes for documentation links to configuring via environment variables.

abadger commented on lib/ansible/utils/module_docs_fragments/oneview.py:
```diff
@@ -0,0 +1,61 @@
+       U(https://github.com/HewlettPackard/oneview-ansible/tree/master/examples)"
+    '''
+
+    VALIDATEETAG = '''
```
Does this and the FACTSPARAMS variables do anything?  I was unaware that the documentation generator would make use of anything besides DOCUMENTATION.

fgbulsoni replied:
Hello, it is apparently able to use it when extending documentation.

I see this inside the FCNetworkModule:
```
extends_documentation_fragment:
    - oneview
    - oneview.validateetag
```

And looking at the `oneview_fc_network_module.rst` file inside `docs`:

```
(...)
        truefalse
        When the ETag Validation is enabled, the request will be conditionally processed only if the current ETag for the resource matches the ETag provided in the data.        
(...)
```
And we also have facts modules which call the FACTSPARAMS for extending documentation, so these should be useful.

fgbulsoni commented:
Many thanks for the review @abadger , will start working on the changes.

fgbulsoni committed:
Changed dual underscore convention for single underscore in all private methods, changed documentation to reflect sphinx more correctly, removed logger, fixed many point commented on the review

fgbulsoni committed:
Adding documentation mentioning to check notes section when trying to setup env vars

fgbulsoni committed:
Adding __metadata__ and changing __future__ import to try to fix code according to boilerplate test

fgbulsoni committed:
Moved ResourceComparator methods to OneViewModuleBase and changed logger to AnsibleModule default .log method

fgbulsoni committed:
Moved python requirement from module to doc fragments, removed deepcopies in comparissons, added set_scope action to fc_networks, made OneViewBaseModule an abc class

fgbulsoni commented:
@abadger I *believe* I've made all of the requested changes. 
Would you perhaps be able to review the PR again?

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,441 @@
+    MSG_ALREADY_PRESENT = 'Resource is already present.'
+    MSG_ALREADY_ABSENT = 'Resource is already absent.'
+    HPE_ONEVIEW_SDK_REQUIRED = 'HPE OneView Python SDK is required for this module.'
+    FUTURE_PACKAGE_REQUIRED = 'The Future Python package is required for this module.'
```
We don't need this message anymore as the future package is no longer used.

fgbulsoni replied:
Ops. True, removing.

fgbulsoni committed:
Removed future required message since future is no longer used

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,441 @@
+
+from ansible.module_utils import six
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils._text import to_native
```
style wise, the ansible imports go below the hpOneView imports.

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,440 @@
+import abc
+import collections
+import json
+import logging
```
We don't need the logging import anymore

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,440 @@
+
+
+class OneViewModuleBase(object):
+    __metaclass__ = abc.ABCMeta
```
Oh sorry, I should have mentioned, how to specify metaclasses changed between python2 and python3.  So there's a helper in six so that you can do this acroos both versions.
```
from ansible.module_utils.six import with_metaclass
[...]
class OneViewModuleBase(with_metaclass(ABCMeta, object)):
```

fgbulsoni replied:
Is it fine to do it as:

```python
@six.add_metaclass(abc.ABCMeta)
class OneViewModuleBase(object):
    MSG_CREATED = 'Resource created successfully.'
    MSG_UPDATED = 'Resource updated successfully.'
```
?
Or would that not work? I've seen it used similarly at the `crypto` module.

abadger commented on lib/ansible/modules/remote_management/hpe/oneview_fc_network.py:
```diff
@@ -0,0 +1,138 @@
+        else:
+            return self.resource_absent(resource)
+
+    def __present(self, resource):
```
There's still a few of these that are using double-underscore to mark private.  We generally do that with single underscore instead.

``` python
- def __present(self,  resource):
+ def _present(self,  resource):
```

fgbulsoni replied:
Ops. Missed that, was a new method that I added and ended up not verifying the underscores. Fixing.

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,440 @@
+            # If both values are null, empty or False it will be considered equal.
+            elif not resource1[key] and not resource2[key]:
+                continue
+            elif isinstance(resource1[key], dict):
```
Use collections.Mapping for dict

fgbulsoni committed:
fixed imports order, removed logging import, changed style of add abcmeta class, removed extra underscores on methods and swapped isinstance validations of dict for collections.Mapping

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,438 @@
+import json
+import os
+import traceback
+from copy import deepcopy
```
No longer need deepcopy

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,438 @@
+
+        :return: dict: It must return a dictionary with the attributes for the module result,
+            such as ansible_facts, msg and changed.
+"""
```
indentation of the end of the string should match with the beginning  (I see this in a few other places as well)

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,438 @@
+            ansible_facts={fact_name: resource}
+        )
+
+    MSG_DIFF_AT_KEY = 'Difference found at key \'{0}\'. '
```
Should this be at the top with the other class attributes?

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,438 @@
+                List with all the scope URIs to be added to the resource.
+        Returns:
+            A dictionary with the expected arguments for the AnsibleModule.exit_json
+    """
```
sphinx docstring formatting was missed for this one

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,438 @@
+                merged_items[item_key] = items_map[item_key].copy()
+                merged_items[item_key].update(item)
+            else:
+                merged_items[item_key] = item.copy()
```
I see that we're still doing copies of the item here twice, once when we enter it into items map at the beginning of hte function and once when we enter it into merged_items here.  Is that needed?

fgbulsoni committed:
Fixed docstrings and docstrings identation, removed a couple of copy methods from merge_lists and moved MSG that was in the middle of the code to top of the class

fgbulsoni commented on lib/ansible/modules/remote_management/hpe/oneview_fc_network.py:
```diff
@@ -0,0 +1,121 @@
+    data:
+        description:
+            - List with the Fibre Channel Network properties.
+        required: true
```
Nice, will probably eventually look into that.

fgbulsoni committed:
Changed condition of not <reference> to is None according to review.

abadger commented on lib/ansible/module_utils/oneview.py:
```diff
@@ -0,0 +1,431 @@
+
+        return str(value)
+
+    def merge_list_by_key(self, original_list, updated_list, key, ignore_when_null=[]):
```
Looks like merge_list_by_key(), _standardize_value(), and _str_sorted() should be moved to be toplevel functions?  (They don't make use of self).

abadger commented on lib/ansible/utils/module_docs_fragments/oneview.py:
```diff
@@ -0,0 +1,65 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
```
If you want to, you can change this license header to the short, one line version as shown here: https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/dev_guide/developing_modules_documenting.rst#copyright

I've been trying to move modules to that since modules are copied over the network to the remote machine and so making them smaller can speed that up.

fgbulsoni replied:
Ok, so, in that link I see this short licence:
```
#!/usr/bin/python
# Copyright (c) 2017 [New Contributor(s)]
# Copyright (c) 2015 [Original Contributor(s)]
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```
I'm assuming the oneview_fc_network.py will then have:
```
#!/usr/bin/python
# Copyright (c) 2016-2017 Hewlett Packard Enterprise Development LP 
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt) 
```
Correct? Also, since that licence is GNU and module_utils must be BSD, is there a similar model for BSD?

fgbulsoni committed:
Moving toplevel functions out of OneViewBaseClass and changing the license header in oneview_fc_network.py to short version

abadger commented:
Your gplv3 question looks correct.  There is none for bsd at this time so
you can leave that one alone.

fgbulsoni commented:
> Your gplv3 question looks correct. There is none for bsd at this time so you can leave that one alone.

Awesome, thanks. Changes made then, waiting on CI to complete to see if I messed up anywhere.

abadger commented:
Merged to devel for 2.4.0.  Thanks for the persistence and hard work!

fgbulsoni commented:
For history purposes, related issue: #28354
